pkg/upgrade/migrate.go

```diff
@@ -0,0 +1,341 @@
+		logCtxt := log.WithField("file", f.Name())
+		logCtxt.Info("processing file...")
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") {
+			logCtxt.Info("skipping and removing last reserved ip file...")
+			if err = os.Remove(fname); err != nil {
+				return fmt.Errorf("failed to remove file %s: %s", fname, err)
+			}
+			logCtxt.Info("successfully removed last reserved ip file!")
+			continue
+		}
+
+		// Skip the lock
+		if f.Name() == "lock" {
+			logCtxt.Info("skipping the lock...")
+			continue
+		}
+
+		// The name of the file is its IP address.
+		ip, _, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name()))
+		if err != nil {
+			return fmt.Errorf("failed to parse IP %s: %s", f.Name(), err)
+		}
+
+		// The contents are the container ID.
+		b, err := ioutil.ReadFile(fname)
+		if err != nil {
+			return fmt.Errorf("failed to read file %s: %s", fname, err)
+		}
+		containerID := string(b)
+
+		// See if the IP hasn't already been assigned for the handleID.
+		handleID := utils.GetHandleID("k8s-pod-network", containerID, "")
+		ips, err := c.IPAM().IPsByHandle(ctxt, handleID)
+		if err == nil {
+		} else if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			// Return error if it is NOT does not exist, otherwise continue.
+			return fmt.Errorf("failed to fetch ips for handleID %s: %s", handleID, err)
+		}
+		logCtxt.WithField("handleID", handleID).Info("file contains handle ID")
+
+		alreadyExists := false
+		for _, ip := range ips {
+			if _, ok := podIPMap[ip.String()]; ok {
+				alreadyExists = true
+				break
+			}
+		}
+		if alreadyExists {
+			logCtxt.Info("pod IP already assigned, skipping...")
+			continue
+		}
+
+		// Get the pod resource associated with the IP.
+		pod, ok := podIPMap[f.Name()]
+		if !ok {
+			logCtxt.WithField("ip", f.Name()).Info("pod not found for IP, deleting and continuing...")
+			if err = os.Remove(fname); err != nil {
+				return fmt.Errorf("failed to remove file %s: %s", fname, err)
+			}
+			continue
+		}
+
+		// Store allocation to datastore
+		logCtxt.Info("assinging pod IP to Calico IPAM...")
+		if err = c.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+			IP:       *ip,
+			HandleID: &handleID,
+			Hostname: nodename,
+			Attrs: map[string]string{
+				ipam.AttributeNode:      nodename,
+				ipam.AttributePod:       pod.Name,
+				ipam.AttributeNamespace: pod.Namespace,
+			},
+		}); err != nil {
+			return fmt.Errorf("failed to assign IP to calico backend: %s", err)
+		}
+		logCtxt.Info("successfully assinged pod IP to Calico IPAM...")
+
+		// Delete the file
+		logCtxt.Info("removing file...")
+		if err = os.Remove(fname); err != nil {
+			return fmt.Errorf("failed to remove file %s: %s", fname, err)
+		}
+		logCtxt.Info("successfully removed file!")
+	}
+
+	// Delete the host-local IPAM data directory
+	log.Info("removing host-local IPAM data directory...")
+	if err = os.Remove(ipAllocPath); err != nil {
```

Comment:
Will this succeed without us deleting the lock file? I think we need to release the lock and delete the file right before this.